### PR TITLE
Support managing superusers

### PR DESCRIPTION
**Pull request by rakeshsharma14317**

### CODE REVIEW

File: cli/pkg/cmdutil/cmdutil.go [hunk 1]
```diff
@@ -202,6 +202,15 @@ func WarnPrinter(str string) {
 	boldYellow.Fprintln(color.Output, str)
 }
 
+func PrintSuperUsers(users []*adminv1.User) {
```
Comment (begelundmuller) [inline cli/pkg/cmdutil/cmdutil.go:205]: Just `PrintUsers`, can be used to render any `User` object

File: cli/cmd/sudo/superuser/remove.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+	"github.com/rilldata/rill/cli/pkg/config"
+	adminv1 "github.com/rilldata/rill/proto/gen/rill/admin/v1"
+	"github.com/spf13/cobra"
+)
+
+func RemoveCmd(cfg *config.Config) *cobra.Command {
+	var email string
+	removeCmd := &cobra.Command{
+		Use:   "remove <email>",
+		Args:  cobra.MaximumNArgs(1),
```
Comment (begelundmuller) [inline cli/cmd/sudo/superuser/remove.go:16]: Must be exact

File: cli/cmd/sudo/superuser/remove.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+			client, err := cmdutil.Client(cfg)
+			if err != nil {
+				return err
+			}
+			defer client.Close()
+
+			if len(args) > 0 {
+				email = args[0]
+			}
```
Comment (begelundmuller) [inline cli/cmd/sudo/superuser/remove.go:29]: Can assume exact 1 arg

File: cli/cmd/sudo/superuser/remove.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+			cmdutil.WarnPrinter(fmt.Sprintf("Warn: Super user removing will revoke all superuser access for %q", email))
+			if !cmdutil.ConfirmPrompt("Do you want to continue", "", false) {
+				cmdutil.WarnPrinter("Aborted")
+				return nil
+			}
```
Comment (begelundmuller) [inline cli/cmd/sudo/superuser/remove.go:35]: No need for a warning

File: cli/cmd/sudo/superuser/remove.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+			cmdutil.SuccessPrinter(fmt.Sprintf("Removed superuser role for %q", email))
+
+			return nil
+		},
+	}
+
+	removeCmd.Flags().SortFlags = false
+	removeCmd.Flags().StringVar(&email, "email", "", "Superuser Email")
```
Comment (begelundmuller) [inline cli/cmd/sudo/superuser/remove.go:51]: Remove, just passing as positional is fine

File: cli/cmd/sudo/superuser/add.go [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+			if len(args) > 0 {
+				email = args[0]
+			}
+
+			cmdutil.WarnPrinter(fmt.Sprintf("Warn: Super user inviting will give all access to the %q", email))
+			if !cmdutil.ConfirmPrompt("Do you want to continue", "", false) {
+				cmdutil.WarnPrinter("Aborted")
+				return nil
+			}
```
Comment (begelundmuller) [inline cli/cmd/sudo/superuser/add.go:35]: No need for warning

File: cli/cmd/sudo/superuser/add.go [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+	"github.com/rilldata/rill/cli/pkg/config"
+	adminv1 "github.com/rilldata/rill/proto/gen/rill/admin/v1"
+	"github.com/spf13/cobra"
+)
+
+func AddCmd(cfg *config.Config) *cobra.Command {
+	var email string
+	addCmd := &cobra.Command{
+		Use:   "add <email>",
+		Args:  cobra.MaximumNArgs(1),
```
Comment (begelundmuller) [inline cli/cmd/sudo/superuser/add.go:16]: Should be exact

File: cli/cmd/sudo/superuser/add.go [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+			client, err := cmdutil.Client(cfg)
+			if err != nil {
+				return err
+			}
+			defer client.Close()
+
+			if len(args) > 0 {
+				email = args[0]
+			}
```
Comment (begelundmuller) [inline cli/cmd/sudo/superuser/add.go:29]: Assume it's exact

File: admin/server/users.go [hunk 9]
```diff
@@ -12,6 +13,72 @@ import (
 	"google.golang.org/protobuf/types/known/timestamppb"
 )
 
+func (s *Server) ListSuperUsers(ctx context.Context, req *adminv1.ListSuperUsersRequest) (*adminv1.ListSuperUsersResponse, error) {
+	claims := auth.GetClaims(ctx)
+	if !claims.Superuser(ctx) {
+		return nil, status.Error(codes.PermissionDenied, "only superusers can add autoinvite domain")
+	}
```
Comment (begelundmuller) [inline admin/server/users.go:20]: error message doesn't match func signature

File: admin/server/users.go [hunk 10]
```diff
@@ -12,6 +13,72 @@ import (
...
+
+	return &adminv1.ListSuperUsersResponse{Users: dtos}, nil
+}
+
+func (s *Server) AddSuperUser(ctx context.Context, req *adminv1.AddSuperUserRequest) (*adminv1.AddSuperUserResponse, error) {
+	claims := auth.GetClaims(ctx)
+	if !claims.Superuser(ctx) {
+		return nil, status.Error(codes.PermissionDenied, "only superusers can add autoinvite domain")
+	}
```
Comment (begelundmuller) [inline admin/server/users.go:42]: error message doesn't match func signature

File: admin/server/users.go [hunk 11]
```diff
@@ -12,6 +13,72 @@ import (
...
+
+	return &adminv1.AddSuperUserResponse{}, nil
+}
+
+func (s *Server) RemoveSuperUser(ctx context.Context, req *adminv1.RemoveSuperUserRequest) (*adminv1.RemoveSuperUserResponse, error) {
+	claims := auth.GetClaims(ctx)
+	if !claims.Superuser(ctx) {
+		return nil, status.Error(codes.PermissionDenied, "only superusers can add autoinvite domain")
+	}
```
Comment (begelundmuller) [inline admin/server/users.go:64]: error message doesn't match func signature

File: admin/server/users.go [hunk 12]
```diff
@@ -12,6 +13,72 @@ import (
...
+
+	users, err := s.admin.DB.FindSuperUsers(ctx)
+	if err != nil {
+		if errors.Is(err, database.ErrNotFound) {
+			return nil, status.Errorf(codes.NotFound, "super user not found")
+		}
```
Comment (begelundmuller) [inline admin/server/users.go:26]: `ErrNotFound` check not needed when listing multiple (not possible to get it)

File: admin/database/database.go [hunk 13]
```diff
@@ -88,6 +88,11 @@ type DB interface {
 
 	ResolveRuntimeSlotsUsed(ctx context.Context) ([]*RuntimeSlotsUsed, error)
 
+	CheckUsersEmpty(ctx context.Context) (bool, error)
+	FindSuperUsers(ctx context.Context) ([]*User, error)
+	AddSuperUser(ctx context.Context, id string) error
+	RemoveSuperUser(ctx context.Context, id string) error
```
Comment (begelundmuller) [inline admin/database/database.go:94]: Just `SetSuperuser(ctx, userID string, superuser bool) error`

Comment (begelundmuller) [inline admin/database/database.go:94]: Move these to the section for `User` model related calls

**Review by begelundmuller — CHANGES_REQUESTED** [review]

File: cli/cmd/sudo/superuser/remove.go (hunk 2 above)
Comment (rakeshsharma14317) [reply cli/cmd/sudo/superuser/remove.go:16]: Do we need this with flags? i mean `--email` ? though of adding that. If not we can have it as exact.

File: admin/database/database.go [hunk 14]
```diff
@@ -88,17 +88,15 @@ type DB interface {
...
-
 	FindUsers(ctx context.Context) ([]*User, error)
 	FindUser(ctx context.Context, id string) (*User, error)
 	FindUserByEmail(ctx context.Context, email string) (*User, error)
 	InsertUser(ctx context.Context, opts *InsertUserOptions) (*User, error)
 	DeleteUser(ctx context.Context, id string) error
 	UpdateUser(ctx context.Context, id string, opts *UpdateUserOptions) (*User, error)
+	CheckUsersEmpty(ctx context.Context) (bool, error)
+	FindSuperUsers(ctx context.Context) ([]*User, error)
+	SetSuperuser(ctx context.Context, userID string, superuser bool) error
```
Comment (begelundmuller) [inline admin/database/database.go:99]: Probably make it `UpdateSuperuser` for consistency with other update calls (only in `database`, `SetSuperuser` is good for the RPC name)

File: admin/database/postgres/postgres.go [hunk 15]
```diff
@@ -739,13 +739,8 @@ func (c *connection) FindSuperUsers(ctx context.Context) ([]*database.User, erro
...
-func (c *connection) AddSuperUser(ctx context.Context, id string) error {
-	res, err := c.getDB(ctx).ExecContext(ctx, `UPDATE users SET superuser=true, updated_on=now() WHERE id=$1`, id)
-	return checkUpdateRow("superuser invite", res, err)
-}
-
-func (c *connection) RemoveSuperUser(ctx context.Context, id string) error {
-	res, err := c.getDB(ctx).ExecContext(ctx, `UPDATE users SET superuser=false, updated_on=now() WHERE id=$1`, id)
+func (c *connection) SetSuperuser(ctx context.Context, userID string, superuser bool) error {
+	res, err := c.getDB(ctx).ExecContext(ctx, `UPDATE users SET superuser=$2, updated_on=now() WHERE id=$1`, userID, superuser)
 	return checkUpdateRow("superuser invite", res, err)
```
Comment (begelundmuller) [inline admin/database/postgres/postgres.go:744]: Should just be `"superuser"` not have `invite`

**Review by begelundmuller — CHANGES_REQUESTED** [review]
Please also change `SuperUser` to `Superuser` everywhere (one word, not two)

**Review by begelundmuller — APPROVED** [review]
Looks good!

**Pull request merged**